add streaming support

shiftkey opened the pull request:
Fixes #101 

Related: https://github.com/desktop/desktop/issues/1818 and https://github.com/desktop/desktop/issues/1524

This PR adds a `GitProcess.spawn` method that has some similar setup to `GitProcess.exec` but requires the caller to interact with the process directly. So while you no longer have the error handling or parsing of `GitProcess.exec`, you have the ability to stream the output directly.

For example, this is how I'm able to workaround the encoding issue from Desktop:

```ts
const path = '/Users/shiftkey/src/spidy'
const args = [ 'diff', 'HEAD', '--no-ext-diff', '--patch-with-raw', '-z', '--no-color', '--binary', '--', 'crawler_done.txt' ]

const stdout: Array<Buffer> = []

const options = { processCallback: setBinaryEncoding }
const process =  GitProcess.spawn(args, path, options)

process.stdout.on('data', (chunk) => {
  // because we set the encoding on `process.stdout` this will always return `Buffer`
  stdout.push(chunk)
})

process.on('exit', (code, signal) => {
  if (code !== 0) {
    // we no longer get the benefit of GitProcess.exec's error code lookups :(
    throw new Error(`exited with code ${code}`)
  }

  const output = Buffer.concat(stdout)
  let encodedText: string | null = null

  try {
    encodedText = output.toString('utf-8')
  } catch (err) {
    // we can handle this error now :tada:
    return
  }

  // or if it works, we just keep on keeping on
})
```

TODO: 
 - [x] a first pass at the API
 - [x] create test repository to demonstrate encoding issue
 - [x] some tests for known good and known error scenarios
 - [x] pull this into Desktop and hammer on the API using our diff problem(s)
 - [x] confirm both related issues are fixed here
 - [x] documentation and guidance for `GitProcess.spawn`

shiftkey commented:
cc @eeejay @BinaryMuse @kuychaco for API/feature feedback

eeejay commented:
This looks like a good API to me. I was trying to figure out how to have it be similar to the Promise API you have with exec, but I think it is nice to just get a ChildProcess back from spawn to work with.

Another interesting case is using stdin to interact with git. For example, wrap `git add -p`.

shiftkey commented:
@eeejay 

> Another interesting case is using stdin to interact with git. For example, wrap `git add -p`.

Yep, our `stdin` story is lacking - I've added our existing `stdin` and `stdinEncoding` but given the ways you can poke at the process after it's started I think I'll drop those.

eeejay commented on lib/git-process.ts:
```diff
@@ -83,6 +119,40 @@ export class GitProcess {
+    }
+
+    if (options && options.processCallback) {
+      options.processCallback(spawnedProcess)
```
You already return the ProcessChild in the function, what does the callback do?

shiftkey replied:
I initially wanted this to be similar to `GitProcess.exec` which lets you manipulate the child process before you then `await` the result. But given we don't have that constraint here we can probably :fire: this too.

shiftkey commented:
Gonna let this sit for a bit today but I'm reasonably happy as a first cut of the API that this addresses the issues we've seen in Desktop around the large diff and string length constraints.

shiftkey commented:
Actually, I'm impatient and want to get this into Desktop.

![](https://cloud.githubusercontent.com/assets/359239/26787048/084b0068-49df-11e7-8f4f-695dee1c3784.gif)

eeejay commented:
\o/

When can we expect a new release with this?

shiftkey commented:
@eeejay it's already available in `1.32.0` :tada:
